### ISSUE TITLE
Keyless signature are invalid for linux/amd64 since v0.1.0

### ISSUE THREAD

**Issue by nicholasdille** [sigstore/gitsign#202]
**Description**

I am trying to verify my installation of gitsign using the keyless signature. Unfortunately, all signatures since v0.1.0 (all non-prerelease versions) fail to verify because the commit SHA in the signature does not match commit SHA of the release tag.

Am I doing something wrong? The same approach works successfully for cosign and rekor.

I have used the following snippets to reproduce this:

1. Choose version to verify:

    ```bash
    VERSION=0.4.1
    ```

1. Download release binaries for gitsign on linux/amd64:

    ```
    for EXT in "" .pem .sig; do
        curl -sLfO https://github.com/sigstore/gitsign/releases/download/v${VERSION}/gitsign_${VERSION}_linux_amd64${EXT}
    done
    ```

1. Fetch commit sha of release tag:

    ```bash
    SHA="$(git ls-remote https://github.com/sigstore/gitsign refs/tags/v${VERSION} | cut -f1)"
    echo "Git SHA for tag v${VERSION}: ${SHA}"
    ```

1. Verify signature:

    ```bash
    COSIGN_EXPERIMENTAL=1 cosign verify-blob gitsign_${VERSION}_linux_amd64 \
        --certificate gitsign_${VERSION}_linux_amd64.pem \
        --signature gitsign_${VERSION}_linux_amd64.sig \
        --certificate-oidc-issuer https://token.actions.githubusercontent.com \
        --certificate-identity "https://github.com/sigstore/gitsign/.github/workflows/release.yml@refs/tags/v${VERSION}" \
        --certificate-github-workflow-name release \
        --certificate-github-workflow-ref "refs/tags/v${VERSION}" \
        --certificate-github-workflow-repository "sigstore/gitsign" \
        --certificate-github-workflow-sha "${SHA}" \
        --certificate-github-workflow-trigger push
    ```

    This fails with the following message:

    ```plaintext
    Error: verifying blob [gitsign_0.4.1_linux_amd64]: validating certRef: expected GitHub Workflow SHA not found in certificate
    main.go:62: error during command execution: verifying blob [gitsign_0.4.1_linux_amd64]: validating certRef: expected GitHub Workflow SHA not found in certificate
    ```

1. Fetch commit SHA in signature:

    ```bash
    curl -sLf https://github.com/sigstore/gitsign/releases/download/v${VERSION}/checksums.txt \
    | grep gitsign_ | grep "linux_amd64$" | cut -d' ' -f1 \
    | xargs rekor-cli search --sha \
    | xargs -I{} rekor-cli get --uuid {} --format json \
    | jq -r '.Body.HashedRekordObj.signature.publicKey.content' \
    | base64 -d \
    | openssl x509 -noout -text
    ```

**Version**

First version: 0.1.0
Last version: 0.4.1

More versions:

```bash
$ cosign version
  ______   ______        _______. __    _______ .__   __.
 /      | /  __  \      /       ||  |  /  _____||  \ |  |
|  ,----'|  |  |  |    |   (----`|  | |  |  __  |   \|  |
|  |     |  |  |  |     \   \    |  | |  | |_ | |  . `  |
|  `----.|  `--'  | .----)   |   |  | |  |__| | |  |\   |
 \______| \______/  |_______/    |__|  \______| |__| \__|
cosign: A tool for Container Signing, Verification and Storage in an OCI registry.

GitVersion:    v1.13.1
GitCommit:     d1c6336475b4be26bb7fb52d97f56ea0a1767f9f
GitTreeState:  clean
BuildDate:     2022-10-17T18:00:05Z
GoVersion:     go1.19.2
Compiler:      gc
Platform:      linux/amd64

$ rekor-cli version
  ____    _____   _  __   ___    ____             ____   _       ___
 |  _ \  | ____| | |/ /  / _ \  |  _ \           / ___| | |     |_ _|
 | |_) | |  _|   | ' /  | | | | | |_) |  _____  | |     | |      | |
 |  _ <  | |___  | . \  | |_| | |  _ <  |_____| | |___  | |___   | |
 |_| \_\ |_____| |_|\_\  \___/  |_| \_\          \____| |_____| |___|
rekor-cli: Rekor CLI

GitVersion:    v1.0.1
GitCommit:     d3162350e96098ca8a24adfdbee42057e43b5de6
GitTreeState:  clean
BuildDate:     2022-11-10T15:26:56Z
GoVersion:     go1.19.3
Compiler:      gc
Platform:      linux/amd64
```

**Comment by wlynch** [comment]
tl;dr you need this for gitsign:

```
SHA="$(git ls-remote https://github.com/sigstore/gitsign refs/tags/v${VERSION}^{} | cut -f1)"
```

Longer answer:

So I think everything is working as intended w.r.t. gitsign releases - what's happening is GitHub Actions isn't including the SHA you expect in it's OIDC claims (which is what is included in the cert), which is why the command doesn't work as-is.

The reason for this is because of [annotated tags](https://git-scm.com/book/en/v2/Git-Basics-Tagging#_annotated_tags). By default when you create a tag in Git, it simply creates a ref pointing from the tag name to the commit SHA you want to tag. The limitation with this is if you only have a simple key value pair you don't have anywhere to put data like tag messages, or more importantly - **tag signatures**.

To work around this, git creates a commit-like object that is used to store the data for the tag, then point that object to the underlying commit. For example for gitsign v0.4.1:

```sh
$ cat .git/refs/tags/v0.4.1
80b829cf31876192707b00ac6441bd2e716a347e          # <--- Value of the tag - this isn't actually a commit
$ git cat-file tag 80b829cf31876192707b00ac6441bd2e716a347e
object 4db581a015a51cfc44cef55feacd2f5ca101c702   # <--- Actual commit value
type commit
tag v0.4.1
tagger Billy Lynch <billy@chainguard.dev> 1669825374 -0500

v0.4.1
-----BEGIN SIGNED MESSAGE-----
MIIEAwYJKoZIhvcNAQcCoIID9DCCA/ACAQExDTALBglghkgBZQMEAgEwCwYJKoZI
hvcNAQcBoIICpTCCAqEwggInoAMCAQICFF1NDqtCGRPUUz+MXxoYZXavzQ3FMAoG
CCqGSM49BAMDMDcxFTATBgNVBAoTDHNpZ3N0b3JlLmRldjEeMBwGA1UEAxMVc2ln
c3RvcmUtaW50ZXJtZWRpYXRlMB4XDTIyMTEzMDE2MjMwMFoXDTIyMTEzMDE2MzMw
MFowADBZMBMGByqGSM49AgEGCCqGSM49AwEHA0IABAcs/MyApzOpDuCGXmlkpRLO
VuLabwzWA59HYAvpuiSWjVE0Nl5kNw24QFRkXg3sVs956j5au9RITu1AXKOqAh2j
ggFGMIIBQjAOBgNVHQ8BAf8EBAMCB4AwEwYDVR0lBAwwCgYIKwYBBQUHAwMwHQYD
VR0OBBYEFAS+FBiymKT6MH2jD5q3nbwv3NvzMB8GA1UdIwQYMBaAFN/T6c9WJBGW
+ajY6ShVosYuGGQ/MCIGA1UdEQEB/wQYMBaBFGJpbGx5QGNoYWluZ3VhcmQuZGV2
MCkGCisGAQQBg78wAQEEG2h0dHBzOi8vYWNjb3VudHMuZ29vZ2xlLmNvbTCBiwYK
KwYBBAHWeQIEAgR9BHsAeQB3AN09MGrGxxEyYxkeHJlnNwKiSl643jyt/4eKcoAv
Ke6OAAABhMlZP5QAAAQDAEgwRgIhAPO+nMENPEiOfb0q/GytnMFp3WIx3PI5Mf5V
N2axXaIxAiEAu9W3MXMJHuE5nKDL/CbOO7MGAS21k4C/udCd7nJpyFQwCgYIKoZI
zj0EAwMDaAAwZQIwV1XHLmop5Mbn6B8EGoFoIwfo8LmPXYcLWr5t3RBxJjwpljBp
wcbrDhyhm9HPrh4CAjEAhaZ03AJr1FwKGswf7StPQhBKnFImQv2sqaYWm3mqVawU
olghFoVCty+kJkWnuaZ0MYIBJDCCASACAQEwTzA3MRUwEwYDVQQKEwxzaWdzdG9y
ZS5kZXYxHjAcBgNVBAMTFXNpZ3N0b3JlLWludGVybWVkaWF0ZQIUXU0Oq0IZE9RT
P4xfGhhldq/NDcUwCwYJYIZIAWUDBAIBoGkwGAYJKoZIhvcNAQkDMQsGCSqGSIb3
DQEHATAcBgkqhkiG9w0BCQUxDxcNMjIxMTMwMTYyMzAwWjAvBgkqhkiG9w0BCQQx
IgQg2WCDUWJrN0Y4AaatqH9IVWVCVUoAxYGHxKUPFfbMV7wwCgYIKoZIzj0EAwIE
RjBEAiB0wC7FXNI2Os56x17GSfgMbQCzoD3HeZt5mLEgVIO4aQIgFM4XCP26J2cb
FEzw5crqMERJPtwKB4eAJcj6EzNpLkg=
-----END SIGNED MESSAGE-----
```

The reason why you're seeing different behavior between cosign and gitsign is because gitsign signs it's tags for its releases (which means it uses annotated tags), while cosign only uses simple lightweight tags. 

If we look at a [sample Rekor entry for v0.4.1](https://rekor.tlog.dev/?logIndex=8236593), you can see that GitHub is populating the workflow_sha claim with the value of the deferenced source commit, not the SHA of the tag itself:

```
data:
  Serial Number: '0x6699f824c86f4697552abc6ad8b7200ca474e0cc'
Signature:
  Issuer: O=sigstore.dev, CN=sigstore-intermediate
  Validity:
    Not Before: 6 days ago (2022-12-01T11:51:50-05:00)
    Not After: 6 days ago (2022-12-01T12:01:50-05:00)
  Algorithm:
    name: ECDSA
    namedCurve: P-256
  Subject:
    extraNames:
      items: {}
    asn: []
X509v3 extensions:
  Key Usage (critical):
  - Digital Signature
  Extended Key Usage:
  - Code Signing
  Subject Key Identifier:
  - 28:90:CE:18:29:26:33:19:0C:88:36:9A:F8:83:A3:AA:9C:3A:BD:9E
  Authority Key Identifier:
    keyid: DF:D3:E9:CF:56:24:11:96:F9:A8:D8:E9:28:55:A2:C6:2E:18:64:3F
  Subject Alternative Name (critical):
    url:
    - https://github.com/sigstore/gitsign/.github/workflows/release.yml@refs/tags/v0.4.1
  OIDC Issuer: https://token.actions.githubusercontent.com
  GitHub Workflow Trigger: push
  GitHub Workflow SHA: 4db581a015a51cfc44cef55feacd2f5ca101c702  # <- Derefenced commit
  GitHub Workflow Name: release
  GitHub Workflow Repository: sigstore/gitsign
  GitHub Workflow Ref: refs/tags/v0.4.1
  1.3.6.1.4.1.11129.2.4.2: 04:7a:00:78:00:76:00:dd:3d:30:6a:c6:c7:11:32:63:19:1e:1c:99:67:37:02:a2:4a:5e:b8:de:3c:ad:ff:87:8a:72:80:2f:29:ee:8e:00:00:01:84:ce:9a:03:98:00:00:04:03:00:47:30:45:02:21:00:f8:91:b5:6e:dd:0b:23:1c:fe:87:2a:0a:3d:66:43:6b:f2:de:0e:aa:26:1b:90:67:08:61:28:65:fb:0e:df:1f:02:20:3a:7c:e9:2c:d7:5a:a1:dd:a6:50:9f:8e:2e:55:44:15:ca:36:95:78:3e:54:5d:3f:ad:b2:75:65:12:17:57:8f
```

Up to you if you think this is a feature or a bug - but either way this is current the behavior of GitHub Actions. 😅 

So what you want to do for your workflow is use the dereferenced value of the annotated tag (i.e. the commit SHA of the underlying source) - which you can do by referencing the revision by [`<tag>^{}`](https://git-scm.com/docs/gitrevisions#Documentation/gitrevisions.txt-emltrevgtemegemv0998em). Luckily you can easily see this in the ref list:

```sh
$ git ls-remote https://github.com/sigstore/gitsign | grep v0.4.1
80b829cf31876192707b00ac6441bd2e716a347e        refs/tags/v0.4.1
4db581a015a51cfc44cef55feacd2f5ca101c702        refs/tags/v0.4.1^{}
$ git ls-remote https://github.com/sigstore/gitsign refs/tags/v0.4.1^{} | cut -f1      

4db581a015a51cfc44cef55feacd2f5ca101c702
```

Which you can plug in to your command above to verify:

```sh
$ SHA=$(git ls-remote https://github.com/sigstore/gitsign refs/tags/v0.4.1^{} | cut -f1)
$ COSIGN_EXPERIMENTAL=1 cosign verify-blob gitsign_${VERSION}_linux_amd64 \
    --certificate gitsign_${VERSION}_linux_amd64.pem \
    --signature gitsign_${VERSION}_linux_amd64.sig \
    --certificate-oidc-issuer https://token.actions.githubusercontent.com \
    --certificate-identity "https://github.com/sigstore/gitsign/.github/workflows/release.yml@refs/tags/v${VERSION}" \
    --certificate-github-workflow-name release \
    --certificate-github-workflow-ref "refs/tags/v${VERSION}" \
    --certificate-github-workflow-repository "sigstore/gitsign" \
    --certificate-github-workflow-sha "${SHA}" \
    --certificate-github-workflow-trigger push
tlog entry verified with uuid: a8c1f3eeaac992552d4f9b362fb263a258606e90e554aba5affb4a6c51586330 index: 8236593
Verified OK
```

Hope this helps! Let me know if I can clarify anything.

**Comment by wlynch** [comment]
cc @steiza FYI (might be worth documenting)

**Comment by nicholasdille** [comment]
Thanks @wlynch for the long explanation. This is pure gold! I also learned about annotated tags and how they are stored in git. Thanks a lot for taking the time to clarify this!
